modules/dnn/src/opencl/reorg.cl

```diff
@@ -53,15 +53,18 @@ __kernel void reorg(const int count,
 {
     for (int index = get_global_id(0); index < count; index += get_global_size(0))
     {
-        int k = index / (height * width);
-        int j = (index - (k * height * width)) / width;
-        int i = (index - (k * height * width)) % width;
+        int sample_size = channels*height*width;
+        int b = index/sample_size;
+        int new_index = index%sample_size;
+        int k = new_index / (height * width);
+        int j = (new_index - (k * height * width)) / width;
+        int i = (new_index - (k * height * width)) % width;
         int out_c = channels / (reorgStride*reorgStride);
         int c2 = k % out_c;
         int offset = k / out_c;
         int w2 = i*reorgStride + offset % reorgStride;
         int h2 = j*reorgStride + offset / reorgStride;
         int in_index = w2 + width*reorgStride*(h2 + height*reorgStride*c2);
-        dst[index] = src[in_index];
+        dst[index] = src[b*sample_size + in_index];
```

Comment:
```cpp
dst[index] = src[in_index];
```
works fine for batches.

Reply:
If I change it back to your proposal only the elements of the first sample will distributed into output samples. I do modulo on the thread-index in order to get right width and height values - thus I add processed sample length `b*sample_size` to the in_index.

Reply:
Got it. I think we need to modify our test to catch it because now it is passed.
I think the best way is to generate target outputs using OpenCV for a single image differs from dog416.png. In example, https://github.com/opencv/opencv_extra/blob/3.4/testdata/dnn/street.png .

This way our test can be rewritten in the following way
```cpp
    void testDarknetModel(const std::string& cfg, const std::string& weights,
                          const std::vector<cv::String>& outNames,
                          const std::vector<std::vector<int> >& refClassIds,
                          const std::vector<std::vector<float> >& refConfidences,
                          const std::vector<std::vector<Rect2d> >& refBoxes,
                          double scoreDiff, double iouDiff, float confThreshold = 0.24) { ...}

    void testDarknetModel(const std::string& cfg, const std::string& weights,
                          const std::vector<cv::String>& outNames,
                          const std::vector<int>& refClassIds,
                          const std::vector<float>& refConfidences,
                          const std::vector<Rect2d>& refBoxes,
                          double scoreDiff, double iouDiff, float confThreshold = 0.24)
    {
        testDarknetModel(cfg, weights, outNames,
                         std::vector<std::vector<int> >(1, refClassIds),
                         std::vector<std::vector<float> >(1, refConfidences),
                         std::vector<std::vector<Rect2d> >(1, refBoxes),
                         scoreDiff, iouDiff, confThreshold = 0.24);
    }
```
where batch size if detected from size of reference vectors.

Some test case:
```cpp
    std::vector<cv::String> outNames(1, "detection_out");

    std::vector<std::vector<int> > classIds(2);
    std::vector<std::vector<float> > confidences(2);
    std::vector<std::vector<Rect2d> > boxes(2);
    classIds[0].resize(3);
    confidences[0].resize(3);
    boxes[0].resize(3);
    classIds[0][0] = 6;  confidences[0][0] = 0.750469f; boxes[0][0] = Rect2d(0.577374, 0.127391, 0.325575, 0.173418);  // a car
    classIds[0][1] = 1;  confidences[0][1] = 0.780879f; boxes[0][1] = Rect2d(0.270762, 0.264102, 0.461713, 0.48131); // a bicycle
    classIds[0][2] = 11; confidences[0][2] = 0.901615f; boxes[0][2] = Rect2d(0.1386, 0.338509, 0.282737, 0.60028);  // a dog

    // TODO: new numbers for street.png
    classIds[1][0] = 6;  confidences[1][0] = 0.750469f; boxes[1][0] = Rect2d(0.577374, 0.127391, 0.325575, 0.173418);  // a car
    classIds[1][1] = 1;  confidences[1][1] = 0.780879f; boxes[1][1] = Rect2d(0.270762, 0.264102, 0.461713, 0.48131); // a bicycle
    classIds[1][2] = 11; confidences[1][2] = 0.901615f; boxes[1][2] = Rect2d(0.1386, 0.338509, 0.282737, 0.60028);  // a dog

    double scoreDiff = (target == DNN_TARGET_OPENCL_FP16 || target == DNN_TARGET_MYRIAD) ? 1e-2 : 8e-5;
    double iouDiff = (target == DNN_TARGET_OPENCL_FP16 || target == DNN_TARGET_MYRIAD) ? 0.013 : 3e-5;

    // batch size 1
    testDarknetModel("yolo-voc.cfg", "yolo-voc.weights", outNames,
                     classIds[0], confidences[0], boxes[0], scoreDiff, iouDiff);

    // batch size 2
    testDarknetModel("yolo-voc.cfg", "yolo-voc.weights", outNames,
                     classIds, confidences, boxes, scoreDiff, iouDiff);
```